src/transformers/models/phi/modeling_phi.py

```diff
@@ -318,6 +318,10 @@ def forward(
         key_states = self.k_proj(hidden_states)
         value_states = self.v_proj(hidden_states)
 
+        # Upcast queries and keys to prevent overflow in Phi-2
+        query_states = query_states.to(torch.float32)
+        key_states = key_states.to(torch.float32)
```

Comment:
that' s not super memory efficient, but if that's what was done in the original model I guess it's a must? 
Overflow in KV from using float16 is something that happened with Llama as well, GPTNeo-X forces float32 for ROPE computation which was enough, something we could try.
Only issue for me is it's not 100% backward compatible because we force the type (may be an issue for quantization cc @younesbelkada

Reply:
I definitely agree, but unfortunately ROPE is already forcing fp32 and that did not fix the issue.

The original model had those lines, though, only Phi-2 had such a problem. In the near future, we aim to improve Phi-2 and mitigate it.

Regarding quantization, do you know how people were quantizing it? Were they using the remote code or the transformers' implementation?

Reply:
I tried forcing ROPE to float32 in different ways but had no success. If somebody has an alternative way or less impactful solution, please feel free to jump in!

Reply:
I narrowed the issue to line 367, where the attention weights is being calculated.

So far, the least impactful I could think of was:
```
attn_weights = torch.matmul(query_states.to(torch.float32), key_states.to(torch.float32).transpose(2, 3)) / math.sqrt(self.head_dim)
```

However, this still requires upcasting queries and keys to float32, with the catch that we are not storing these anymore.

It still might be a problem with quantization, right?

Reply:
> `attn_weights = torch.matmul(query_states.to(torch.float32), key_states.to(torch.float32).transpose(2, 3)) / math.sqrt(self.head_dim)`

Sounds great too ! Otherwise we can go for https://github.com/huggingface/transformers/pull/28163/files#r1446995070